tests/admin/views/test_releases_v2.py

```diff
@@ -0,0 +1,1293 @@
+    assert ret.status_code == 400
+
+
+@pytest.mark.usefixtures("releases_db")
+def test_put_fails_without_permission(api, firefox_60_0b3_build1, mock_verified_userinfo):
+    mock_verified_userinfo("notbob")
+    firefox_60_0b3_build1 = deepcopy(firefox_60_0b3_build1)
+    firefox_60_0b3_build1["detailsUrl"] = "https://newurl"
+    firefox_60_0b3_build1["platforms"]["Darwin_x86_64-gcc3-u-i386-x86_64"]["locales"]["de"]["buildID"] = "9999999999999"
+    old_data_versions = populate_versions_dict(firefox_60_0b3_build1)
+
+    ret = api.put("/v2/releases/Firefox-60.0b3-build1", json={"blob": firefox_60_0b3_build1, "product": "Firefox", "old_data_versions": old_data_versions})
+    assert ret.status_code == 403, ret.data
+
+
+@pytest.mark.usefixtures("releases_db")
+def test_put_fails_without_permission_new_release(api, firefox_62_0_build1, mock_verified_userinfo):
+    mock_verified_userinfo("notbob")
+
+    ret = api.put("/v2/releases/Firefox-62.0-build1", json={"blob": firefox_62_0_build1, "product": "Firefox"})
+    assert ret.status_code == 403, ret.data
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_put_fails_for_invalid_release(api):
+    # Only submit a portion of what's needed to have a valid blob
+    data = {"platforms": {"Linux_x86_64-gcc3": {"platforms": {"de": {"appVersion": "65.0", "buildID": "9090909090990", "displayVersion": "65.0"}}}}}
+
+    ret = api.put("/v2/releases/Firefox-65.0-build1", json={"blob": data})
+    assert ret.status_code == 400, ret.data
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_put_fails_for_readonly_release(api, firefox_60_0b3_build1):
+    dbo.releases_json.t.update(values={"read_only": True}).where(dbo.releases_json.name == "Firefox-60.0b3-build1").execute()
+
+    firefox_60_0b3_build1 = deepcopy(firefox_60_0b3_build1)
+    firefox_60_0b3_build1["detailsUrl"] = "https://newurl"
+    firefox_60_0b3_build1["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]["buildID"] = "9999999999999"
+    old_data_versions = populate_versions_dict(firefox_60_0b3_build1)
+
+    ret = api.put("/v2/releases/Firefox-60.0b3-build1", json={"blob": firefox_60_0b3_build1, "product": "Firefox", "old_data_versions": old_data_versions})
+    assert ret.status_code == 400, ret.data
+    assert "Cannot overwrite" in ret.json["exception"]
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_put_succeeds(api, firefox_60_0b3_build1):
+    firefox_60_0b3_build1 = deepcopy(firefox_60_0b3_build1)
+    firefox_60_0b3_build1["detailsUrl"] = "https://newurl"
+    firefox_60_0b3_build1["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]["buildID"] = "9999999999999"
+
+    old_data_versions = populate_versions_dict(firefox_60_0b3_build1)
+    new_data_versions = versions_dict(default=2)
+    assert new_data_versions["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]
```

Comment:
I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.

To get closer to having an explicit expectation while holding onto the convenience of creating that, what about 
* using `versions_dict(default=None)`
* explicitly setting the expected values on the base and the modified locale

I'm hoping that will error out if any error creeps into the api response (ie 2 is not equal to None).

Reply:
> I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.

I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.

> To get closer to having an explicit expectation while holding onto the convenience of creating that, what about
> 
>     * using `versions_dict(default=None)`
>     * explicitly setting the expected values on the base and the modified locale

Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.

Reply:
> > I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.
> 
> I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.

I was thinking of the case where ret.json contains more data than it should and still passes the test, because of an incorrect assumption on my part, but it doesn't appear that this happens . A `versions_dict` will create a new key/default when it's accessed or asserted, but not when doing an equality comparison with `ret.json`. 

> Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.

Do you mean for clarity in the test rather than sorta-obscuring it with the include_base parameter on versions_dict ?

Reply:
> > > I feel uneasy about `new_data_versions` - a full dict of 2's when only a couple of entries are expected to take that value. The assert must be only checking keys in `ret.json`, otherwise it would fail.
> > 
> > 
> > I think I might be understanding your comment, but to be clear: I think `new_data_versions` only contains a `.` entry and one for `["platforms"]["Linux_x86_64-gcc3"]["locales"]["en-US"]` since `versions_dict` doesn't do any population aside from `.`.
> 
> I was thinking of the case where ret.json contains more data than it should and still passes the test, because of an incorrect assumption on my part, but it doesn't appear that this happens . A `versions_dict` will create a new key/default when it's accessed or asserted, but not when doing an equality comparison with `ret.json`.

Ahhhhh, I see. I'm actually a little surprised this is true now that you mention it. I'm guessing that the assert walks through the object with `.keys()`, which wouldn't trigger the defaults...

> 
> > Regardless of the above, I can see it being useful to avoid setting `.` at all in `versions_dict` and having the test methods do that.
> 
> Do you mean for clarity in the test rather than sorta-obscuring it with the include_base parameter on versions_dict ?

Yeah, exactly.